n3fit/src/evolven3fit/utils.py

```diff
@@ -19,45 +19,84 @@ def get_theoryID_from_runcard(usr_path):
     return my_runcard["theory"]["theoryid"]
 
 
-def generate_q2grid(Q0, Qfin, Q_points, match_dict, nf0=None, legacy40=False):
-    """Generate the q2grid used in the final evolved pdfs or use the default grid if Qfin or Q_points is
-    not provided.
+def generate_q2grid(Q0, Qfin, Q_points, match_dict, legacy40=False):
+    """Generate the q2grid used in the final evolved pdfs or use the default grid if legacy40 is set. 
+    The grid uses log(log(Q2/Lambda2)) spacing between the points, with Lambda2=0.0625.
+
+    The grid from Q2_min --> mc is made separately from the rest to be able to let it contain at
+    least 5 points. The rest of the grids contains the same loglog spacing from threshold to threshold.
+
+    Q2_min --> mc^2 --> Q0^2 --> mb^2 --> Q2_max
 
-    match_dict contains the couples (mass : factor) where factor is the number to be multiplied to mass
-    in order to obtain the relative matching scale.
+    match_dict contains the quark mass thresholds and factors. Factor is the number to be 
+    multiplied to mass in order to obtain the relative matching scale.
     """
-    if Qfin is None and Q_points is None:
-        if legacy40:
-            return Q2GRID_NNPDF40
-        elif nf0 in (3, 4, 5):
-            return Q2GRID_DEFAULT
-        elif nf0 is None:
-            raise ValueError("In order to use a default grid, a value of nf0 must be provided")
+    
+    # If flag --legacy40 is set return handmade legacy grid
+    if legacy40:
+        return Q2GRID_NNPDF40
+    # Otherwise dynamically create the grid from Q2_min --> Q2_max
+    else: 
+        if Qfin is None and Q_points is None:
+            Q2_min = 1.0**2
+            Q2_max = 1e5**2
+            Q0 = 1.65                    
+            Lambda2 = 0.0625                  
+            total_points = 50
+            total_points_ic = 6
```

Comment:
as @scarlehoff suggested: make all of those parameters an input to this function - except `Lambda2`, which I suggest to make a module constant instead (since it is unlikely to change) (i.e. also rename it to `LAMBDA2`)